src/identity/storage/WrappedFetchAdapterFactory.ts

```diff
@@ -17,51 +21,95 @@ import namedNode = DataFactory.namedNode;
+      }
+      const data = await response.text();
+      let json: any | undefined;
       try {
-        dataset = await fetchDataset(id);
-      } catch {
-        this.logger.debug(`Looking for payload data failed at ${id}`);
-        return payload;
+        json = JSON.parse(data);
+        // We can only parse as simple JSON if the @context is correct
+        if (json['@context'] !== 'https://www.w3.org/ns/solid/oidc-context.jsonld') {
+          throw new Error('Invalid context');
+        }
+      } catch (error: unknown) {
+        json = undefined;
+        this.logger.debug(`Found unexpected client WebID for ${id}: ${createErrorMessage(error)}`);
       }
 
-      // Get the OIDC Registration JSON
-      const rawRegistrationJsonQuads = dataset.match(namedNode(id), SOLID.terms.oidcRegistration);
-
-      // Check all the registrations to see if any are valid.
-      for (const rawRegistrationJsonQuad of rawRegistrationJsonQuads) {
-        try {
-          return this.validateRegistrationQuad(rawRegistrationJsonQuad, id);
-        } catch {
-          // Keep looking for a valid quad
+      if (json) {
+        // Need to make sure the document is about the id
+        if (json.client_id !== id) {
+          throw new Error('The client registration `client_id` field must match the client WebID');
         }
+        payload = json;
+      } else {
+        // Since the WebID does not match the default JSON-LD we try to interpret it as RDF
+        payload = await this.parseRdfWebId(data, id, response);
       }
-      this.logger.debug(`No payload data was found at ${id}`);
+
+      // `token_endpoint_auth_method: 'none'` prevents oidc-provider from requiring a client_secret
+      payload = { ...payload, token_endpoint_auth_method: 'none' };
     }
 
-    // Will also be returned if no valid registration data was found above
+    // Will also be returned if no valid client data was found above
     return payload;
   }
 
+  private async parseRdfWebId(data: string, id: string, response: Response): Promise<AdapterPayload> {
+    const contentType = response.headers.get('content-type');
+    if (!contentType) {
+      throw new Error(`No content-type received for client WebID ${id}`);
+    }
+
+    // Try to convert to quads
+    const representation = new BasicRepresentation(data, contentType);
+    const preferences = { type: { [INTERNAL_QUADS]: 1 }};
+    const converted = await this.converter.handleSafe({ representation, identifier: { path: id }, preferences });
+    const quads = new Store();
+    const importer = quads.import(converted.data);
+    await new Promise((resolve, reject): void => {
+      importer.on('end', resolve);
+      importer.on('error', reject);
+    });
+
+    // Find the valid redirect uris
+    const match = quads.getObjects(id, 'http://www.w3.org/ns/solid/oidc#redirect_uris', null);
+
+    return {
+      client_id: id,
+      redirect_uris: match.map((node): string => node.value),
```

Comment:
Probably want to filter on valid URLs?

Reply:
If there are invalid URLs there that means there's an issue with the WebID. I don't think that's something we need to solve.

Reply:
Not solve, but just as as security measure, weed out strings that are invalid.

Reply:
From a security point of view there is no issue here. How this works is that a client does an auth request that contains the redirect URI the OIDC library needs to send the response to. Part of the transaction is that the OIDC library compares if that redirect URI is present in the redirect_uris array of the client metadata. This is simply string comparison so no security issues there.

From a semantics point of view this class interprets the data at the given URL as the client metadata (because that's what the spec tells us to do), so if that metadata says a specific string is one of their redirect_uris, it's not really our place to doubt that. 

This same redirect_uris field is also present when we can parse it as a JSON object in the function above. It's not put there explicitly in the code since at that point we can just copy the fields directly due to the `@context` field.